### PR TITLE
Updated reduction packaging to include wavelength ranges

### LINKED ISSUE #21734: SANS GUI V2 needs option to specify wavelength rebin string manually

### PR DESCRIPTION
**Pull request by Matthew-Andrew**
This PR allows a range set of wavelength ranges to be selected in the new ISIS SANS GUI. A reduction will then be done between each of these wavelength ranges with the resulting workspaces being output to the ADS in a group. 

**To test:**

- Get the sample data from here
[loqdemo.zip](https://github.com/mantidproject/mantid/files/1961183/loqdemo.zip)

- In the new ISIS sans GUI
- Load maskfile.txt as the user file
- Load load batch_mode_reduction.csv as the batch file.
- In the bottom right ensure that Use Optimizations is unticked.
- Go to Settings-> Q,Wavelength, Detector Limits
- In the wavelength group box change the step type to RangeLin
- Change the ranges box to 1:1:3
- Go back to the runs tab and click process.
- At the end the ADS should contain 3 group workspaces.
   - first_time_merged should contain 3 workspaces 
   - second_time_merged should contain 3 workspaces
   - LAB_and_HAB_workspaces_from_merged_reduction should contain 12 workspaces
- Check that the names all make sense

Fixes #21734 

. 

**Release Notes** 

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [x] Is the code of an acceptable quality?
- [x] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)?
- [x] Are the unit tests small and test the class in isolation?
- [x] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [x] Do changes function as described? Add comments below that describe the tests performed?
- [ ] Do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by soininen** [comment]
I spotted some layout problems in the settings tab:
sans-settings-glitches

File: scripts/SANS/sans/common/general_functions.py [hunk 1]
```diff
@@ -686,33 +686,55 @@ def get_standard_output_workspace_name(state, reduction_data_type, transmission
...
+        end_time = slice_state.end_time
+        if start_time and end_time:
+            start_time_as_string = '_t%.2f' % start_time[0]
+            end_time_as_string = '_T%.2f' % end_time[0]
+        else:
+            start_time_as_string = ""
+            end_time_as_string = ""
+        user_specified_output_name = user_specified_output_name + start_time_as_string + end_time_as_string
+
+    if multi_reduction_type and multi_reduction_type["wavelength_range"] and user_specified_output_name:
```
Comment (soininen) [inline scripts/SANS/sans/common/general_functions.py]: Could these three `if` statements be simplified to something like:
```python
if multi_reduction_type and user_specified_output_name:
    if multi_reduction_type["period"]:
        pass
    if multi_reduction_type["event_slice"]:
        pass
    if multi_reduction_type["wavelength_range"]:
        pass
```

File: scripts/SANS/sans/gui_logic/models/state_gui_model.py [hunk 2]
```diff
@@ -367,7 +367,7 @@ def event_binning(self, value):
     # - wavelength_and_pixel_adjustment
     # This is not something that needs to be known at this point, but it is good to know.
     # ------------------------------------------------------------------------------------------------------------------
-    def _update_wavelength(self, min_value=None, max_value=None, step=None, step_type=None):
+    def _update_wavelength(self, min_value=None, max_value=None, step=None, step_type=None, range=None):
```
Comment (soininen) [inline scripts/SANS/sans/gui_logic/models/state_gui_model.py]: `range` is a reserved keyword in Python, might be a good idea to consider something else here.

File: scripts/SANS/sans/state/state_base.py [hunk 3]
```diff
@@ -248,6 +254,18 @@ def _type_check(self, value):
...
+class PositiveFloatListParameter(TypedParameter):
+    def __init__(self):
+        super(PositiveFloatListParameter, self).__init__(list, all_list_elements_are_float_and_positive_and_not_empty)
+
+    def _type_check(self, value):
+
+        if not isinstance(value, self.parameter_type) or not all_list_elements_are_float_and_not_empty(value):
+            raise TypeError("Trying to set {0} which expects a value of type {1}."
+                            " Got a value of {2} which is of type: {3}".format(self.name, str(self.parameter_type),
+                                                                               str(value), type(value)))
```
Comment (soininen) [inline scripts/SANS/sans/state/state_base.py]: Are the `str()` conversions needed here? Wouldn't `format()` do this automatically?

**Review by soininen — CHANGES_REQUESTED** [review]
The test procedure described in the description worked flawlessly after I got the data search directories set up correctly. If the reduction fails due to no files found the GUI still seems to go full steam ahead. The only indication of issues is in the logs. Might be worth considering in the future to color the problematic rows with red or orange. Anyway, the code changes look good and clean. Just a few minor comments below.

File: scripts/SANS/sans/algorithm_detail/batch_execution.py [hunk 4]
```diff
@@ -749,7 +818,17 @@ def add_to_group(workspace, name_of_group_workspace):
...
+        if type(group_workspace) is WorkspaceGroup:
+            group_workspace.add(name_of_workspace)
+        else:
+            group_name = "GroupWorkspaces"
+            group_options = {"InputWorkspaces": [name_of_workspace],
+                             "OutputWorkspace": name_of_group_workspace}
+            group_alg = create_unmanaged_algorithm(group_name, **group_options)
+            # At this point we are dealing with the ADS, hence we need to make sure that this is not called as
+            # a child algorithm
+            group_alg.setChild(False)
```
[the remark this answers is not in the document]
Comment (Mantid-Matthew) [reply scripts/SANS/sans/algorithm_detail/batch_execution.py]: Thanks thats good to know. Am I right in thinking that setChild controls the ADS and whether workspace history is stored ? What happens if I setChild to true and setAlwaysStoreInADS to true ?

File: scripts/SANS/sans/algorithm_detail/batch_execution.py [hunk 5]
```diff
@@ -434,6 +434,70 @@ def reduction_packages_require_splitting_for_event_slices(reduction_packages):
     return requires_split
 
 
+def reduction_packages_require_splitting_for_wavelength_range(reduction_packages):
+    """
+        Creates reduction packages from a list of reduction packages by splitting up wavelength ranges.
+
+        The SANSSingleReduction algorithm can handle only a single wavelength range. For each wavelength range, we require an individual
+        reduction. Hence we split the states up at this point.
+        :param reduction_packages: a list of reduction packages.
```
[the remark this answers is not in the document]
Comment (Mantid-Matthew) [reply scripts/SANS/sans/algorithm_detail/batch_execution.py]: is the normal syntax not :param name: ?

File: scripts/SANS/sans/algorithm_detail/batch_execution.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (soininen) [reply scripts/SANS/sans/algorithm_detail/batch_execution.py]: `setChild()` indeed controls the ADS and history, `setAlwaysStoreInADS()` gives a bit more fine-grained control. In the background they control the same flag so setting `setAlwaysStoreInADS(True)` after calling `setChild(True)` will put all workspaces to the ADS.

File: scripts/SANS/sans/algorithm_detail/batch_execution.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (soininen) [reply scripts/SANS/sans/algorithm_detail/batch_execution.py]: Yes, of course, my mistake. Sorry for the noise!

**Comment by soininen** [comment]
I wonder if something went wrong when fixing the glitches since the Wavelength box in the settings view now looks quite differrent; the Ranges field is gone, replaced by Min and Max:
sans-settings-wavelength-missing-ranges

**Comment by Mantid-Matthew** [comment]
The wavelength box should change its appearance depending on which step type you select. For step type Lin and Log what you are seeing is correct however it should change to display a single range box when RangeLin or RangeLog are selected. 

I notice that the Q resolution box is still being cut of for you at the bottom so I'll have another look at that.

**Review by soininen — APPROVED** [review]

**Comment by soininen** [comment]
The missing range field was indeed a problem on my side. The layout is now fixed. I think is good to go. Great work!

**Pull request merged**